### ISSUE TITLE
Behavior change with new sse2 impl of line index

### ISSUE THREAD

**Issue by azdavis** [rust-lang/rust-analyzer#15273]
I made a script which checks that line-index outputs the same `LineIndex` for a given test string, `"\n空"`, for both the old impl (pre-sse2 changes) and the new impl. 

The script is:

```sh
#!/bin/sh

set -eu

tests=lib/line-index/src/tests.rs

check() {
  git checkout -- "$tests"
  git checkout "$1"
  cat <<EOF > "$tests"
#[test]
fn newline_chinese_char() {
  let text = "\n空";
  let idx = crate::LineIndex::new(text);
  println!("{idx:#?}");
}
EOF
  cargo test --lib --package line-index -- newline_chinese_char --nocapture > "$2"
}

# this commit 'Speedup line index calculation via SSE2' adds the new impl
new_impl=97748b0fc6d132504437777644ad3a240c417b14

check "$new_impl~" old
check "$new_impl" new
diff old new
echo 'no diff'
```

You should run it from the root of the repo.

The `diff` fails, indicating that there is a behavioral difference between the old and new impls:

```diff
10,11c10,11
<                 start: 0,
<                 end: 3,
---
>                 start: 1,
>                 end: 4,
```

**Comment by azdavis** [comment]
I realize I said the issue was with the new sse2 impl but I'm actually not sure whether the sse2 code path is getting selected. It could also be the issue is with the new generic non-sse2 impl, or both.

**Comment by lnicola** [comment]
Can you show the full `Debug` output? I assume it's pretty small.

**Comment by azdavis** [comment]
I tested locally by removing the sse2 version and confirmed it's the new generic version which behaves differently. I wasn't able to test the new sse2 version - I tried by changing the new generic version to always panic, and it did panic. Maybe my local setup doesn't have the sse2 feature available. (I'm not sure about that.)

**Comment by azdavis** [comment]
Old:

```
LineIndex {
    newlines: [
        1,
    ],
    line_wide_chars: {
        1: [
            WideChar {
                start: 0,
                end: 3,
            },
        ],
    },
    len: 4,
}
```

New:

```
LineIndex {
    newlines: [
        1,
    ],
    line_wide_chars: {
        1: [
            WideChar {
                start: 1,
                end: 4,
            },
        ],
    },
    len: 4,
}
```

**Comment by lnicola** [comment]
I think the old version stored the wide char position as an [offset](https://github.com/rust-lang/rust-analyzer/commit/44b826afb2091444d54a0762e7784c0dd6dbda26#diff-fc40a85bc2dc8b411627619ea2f899b81b1bbb82cc426d5dbe11a19278ea5954L118-L119) from the start of the line, while the new version [doesn't](https://github.com/rust-lang/rust-analyzer/commit/44b826afb2091444d54a0762e7784c0dd6dbda26#diff-fc40a85bc2dc8b411627619ea2f899b81b1bbb82cc426d5dbe11a19278ea5954R366).

**Comment by azdavis** [comment]
I think this means that the new version gives the wrong answers in some cases, right? Since none of the API method signatures or impls changed.

**Comment by lnicola** [comment]
I don't know if they're wrong or if the meaning of the field changed. Not like it's documented 😅.

**Comment by azdavis** [comment]
They are [documented](https://github.com/rust-lang/rust-analyzer/blob/97748b0fc6d132504437777644ad3a240c417b14/lib/line-index/src/lib.rs/#L54-L60) as offsets, aren't they?

**Comment by azdavis** [comment]
Maybe a solution would be to revert the slow path to the old slow path implementation, and then use that to make sure the fast path has the same behavior as the slow path?

**Comment by lnicola** [comment]
@Veykril I wanted to take a closer look at this but didn't have a chance :-(. Do you remember if the change in meaning of the offsets was intentional?

**Comment by Veykril** [comment]
No it wasn't

**Comment by roife** [comment]
I believe the problem mentioned in this issue has been addressed in #16041.

**Comment by lnicola** [comment]
Heads-up: https://crates.io/crates/line-index/0.1.1 is now out and includes the fix for this issue.
